### PR TITLE
Consistent definition PointLocalCoordiantes between 3D Quad and Triangle + Adding analytical solution for tetrahedra

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

**Comment by loumalouomega** [comment]
I added a test for the local coordinates of a tetrahedra. @rubenzorrilla Can you check it works for you?, this will reduce the number of iterations for the tetrahedra to always 1, I hope this will help you

File: kratos/geometries/triangle_3d_3.h
Comment (pooyan-dadvand) [inline kratos/geometries/triangle_3d_3.h]: A minor comment, the standard way to say that a method is internal is to name it with Implemenation and make it private method: 
```c++
private:
PointLocalCoordinatesImplementation(){ ...}
```

**Review by pooyan-dadvand — COMMENTED** [review]
@rubenzorrilla I use this function many times in distance calculateion, would you please check if it works for you?You may approve it then directly.

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: I used that name because we chose that name for the Quad. Do I change both?

**Comment by loumalouomega** [comment]
The Aussas tests are failing, so I assume that the implementation is not perfect. I added cpp test for the PointLocalCoordinates and the  IsInside methods, but maybe the tests are passing because those are quite basic tests. @rubenzorrilla can you confirm that this methods doesn't work?, I will try to improve the tests

File: kratos/geometries/triangle_3d_3.h
Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_3.h]: It would be better to change them so.

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: Okidoki

**Comment by pooyan-dadvand** [comment]
@loumalouomega it is **crucial to clearly state** in your PR if some tests are not working!! It is not something that I assume by myself.... 

Thanks to @roigcarlo that tests are activated in travis.

**Comment by loumalouomega** [comment]
@pooyan-dadvand I added a test in purpose to check that, I am sorry, I didn't check those tests. In any case I was waiting for the reply of @rubenzorrilla to know what to check, I wasn't assuming It was just right

**Comment by loumalouomega** [comment]
Now it run the tests...except some tests not related to this PR (TestPropertiesHasVariables and TestPropertiesIsEmpty)

**Comment by rubenzorrilla** [comment]
@loumalouomega we did realize today that the Ausas test was not working, so probably this is not related to your implementation. I'm currently trying to figure out why.

@pooyan-dadvand, with regard to the distance computation, I will check it ASAP.

File: kratos/geometries/geometry.h
Comment (RiccardoRossi) [inline kratos/geometries/geometry.h]: this is an error not a warning...

File: kratos/geometries/quadrilateral_3d_4.h
Comment (RiccardoRossi) [inline kratos/geometries/quadrilateral_3d_4.h]: error

Comment (RiccardoRossi) [reply kratos/geometries/quadrilateral_3d_4.h]: or maybe we could pass some way to check if successful...

File: kratos/tests/geometries/test_tetrahedra_3d_4.cpp
Comment (RiccardoRossi) [inline kratos/tests/geometries/test_tetrahedra_3d_4.cpp]: minor:
   
      baricentre /= 3.0;

**Review by RiccardoRossi — COMMENTED** [review]
i think we should throw an error, or at least devise a way to check for success. 

@rubenzorrilla you are using this heavily. Does this suit you?

**Comment by rubenzorrilla** [comment]
@RiccardoRossi I did not deeply check it yet. Please @loumalouomega wait for me to try it.

File: kratos/geometries/quadrilateral_3d_4.h
Comment (rubenzorrilla) [inline kratos/geometries/quadrilateral_3d_4.h:518]: Why did you change the name of the method? IMO PrivatePointLocalCoordinates is more self explanatory.

File: kratos/geometries/triangle_3d_3.h
Comment (rubenzorrilla) [inline kratos/geometries/triangle_3d_3.h:869]: Maybe I'm missing something, but just form my understanding, which is the problem in particularizing the implementation in the PointLocalCoordinates method instead of calling an auxiliar PointLocalCoordinatesImplementation (or PrivatePointLocalCoordinates)?

**Comment by rubenzorrilla** [comment]
As I mentioned before, the issue with the Ausas condition test was not related to @loumalouomega 's modifications and it has been addressed in PR #1659. 

@loumalouomega would you mind to either cherrypick that commit or merge the master once PR #1659 is merged to be sure that everything is fine.

Besides, I ran the core tests, in where the distance processes as well as the modified shape functions and divide utilities are tested (these features widely use the 3D triangle and tetrahedron PointLocalCoordinates method). Everything was as expected,

File: kratos/geometries/geometry.h
Comment (loumalouomega) [reply kratos/geometries/geometry.h]: If I throw an error cases where the point to compute is very far* will stop the simulation.

* (like in contact, where you may be interested in check this kind of cases)

Comment (loumalouomega) [reply kratos/geometries/geometry.h]: This was the motive why the quadrilateral contact was more robust than the triangular contact.

Comment (loumalouomega) [reply kratos/geometries/geometry.h]: It was throwing an  error only in the triangle 3d, in the other classes was using just a std::cout, I did the change for consistency and in my opinion is the proper way

File: kratos/geometries/quadrilateral_3d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: I already said about the error. Besides, see the previous code -> std::cout

File: kratos/geometries/quadrilateral_3d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h:518]: I did the change @pooyan-dadvand  suggested

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h:869]: This was a discussion long time ago, I can not find the issue, but it was discussed that way on that moment

File: kratos/geometries/geometry.h
Comment (loumalouomega) [reply kratos/geometries/geometry.h]: Besides I am using heavily too

**Comment by loumalouomega** [comment]
I can wait, I cherrypicked in a different branch what I am interested in use

File: kratos/geometries/triangle_3d_3.h
Comment (RiccardoRossi) [inline kratos/geometries/triangle_3d_3.h]: J and invJ could be bounded matrices in this function

File: kratos/geometries/triangle_3d_3.h
Comment (RiccardoRossi) [inline kratos/geometries/triangle_3d_3.h]: noalias

**Review by RiccardoRossi — COMMENTED** [review]
minor

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: True

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: True

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

File: kratos/geometries/triangle_3d_3.h
Comment (roigcarlo) [inline kratos/geometries/triangle_3d_3.h]: It would be nice if this check were in the loop, but apart from that ok!

File: kratos/geometries/triangle_3d_3.h
Comment (roigcarlo) [inline kratos/geometries/triangle_3d_3.h]: Same

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: What do you mean?, the check is on the loop

File: kratos/geometries/triangle_3d_3.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: Idem

**Comment by loumalouomega** [comment]
@rubenzorrilla have you tested?

**Comment by loumalouomega** [comment]
@RiccardoRossi can we merge this?, @rubenzorrilla approves

**Comment by loumalouomega** [comment]
@rubenzorrilla

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h:518]: We have used the implementation suffix many places when it is a function body to be used in different methods. The private is already marked by the type of the method (private/public). For classes the new `Internals `namespace (or nested class) should be used

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h]: I like the idea of @RiccardoRossi for returning if is successful 👍

Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Do you mean to print "successful" or to return a boolean?, the first option will emply a lot of verbosity. This is a method quite used

File: kratos/geometries/triangle_3d_3.h
Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_3.h:869]: I'm not sure but I think the problem was the implementation of `IsInside` and not the local coordinate itself.

File: kratos/geometries/geometry.h
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: I would suggest to change the 30 with a constexpr

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h]: To return a bool but this would imply changes in the interface

File: kratos/geometries/geometry.h
Comment (loumalouomega) [reply kratos/geometries/geometry.h]: Do I change it here or in a different PR?

Comment (pooyan-dadvand) [reply kratos/geometries/geometry.h]: Just here would be better

Comment (pooyan-dadvand) [reply kratos/geometries/geometry.h]: Also the rest of the constant numbers

**Review by pooyan-dadvand — APPROVED** [review]
👍

**Comment by loumalouomega** [comment]
Thanks, I will wait for appveyor

**Pull request merged**
